hnn_core/parallel.py

```diff
@@ -7,20 +7,85 @@
 
 from neuron import h
 
-rank = 0
-nhosts = 1
-pc = h.ParallelContext(nhosts)
-pc.done()
-rank = int(pc.id())
-cvode = h.CVode()
+pc = None
+last_network = None
+rank = None
 
 
-def create_parallel_context(n_jobs=1):
-    """Create parallel context."""
-    rank = int(pc.id())     # rank or node number (0 will be the master)
+def shutdown():
+    """Tell nrniv processes to terminate. Calling script will also terminate"""
+    pc.done()
+    h.quit()
 
-    if rank == 0:
-        pc.gid_clear()
+
+def get_nhosts():
+    """Return the number of processors used by ParallelContext
+    Returns
+    -------
+    nhosts: int
+        Value from pc.nhost()
+    """
+    return nhosts
+
+
+def get_rank():
+    """Return the MPI rank from ParallelContext
+    Returns
+    -------
+    rank: int
+        Value from pc.id()
+    """
+    return rank
+
+
+def create_parallel_context(n_cores=None):
+    """Create parallel context.
+    Parameters
+    ----------
+    n_cores: int | None
+        Number of processors to use for a simulation. A value of None will
+        allow NEURON to use all available processors.
+    """
+
+    global rank, nhosts, cvode, pc, last_network
+
+    if pc is None:
+        if n_cores is None:
+            # MPI: Initialize the ParallelContext class
+            pc = h.ParallelContext()
+        else:
+            pc = h.ParallelContext(n_cores)
+
+        nhosts = int(pc.nhost())  # Find number of hosts
+        rank = int(pc.id())     # rank or node number (0 will be the master)
+        cvode = h.CVode()
+
+        # be explicit about using fixed step integration
+        cvode.active(0)
+
+        # use cache_efficient mode for allocating elements in contiguous order
+        # cvode.cache_efficient(1)
+    else:
+        # ParallelContext() has already been called. Don't start more workers.
+        # Just tell old nrniv workers to quit.
+        pc.done()
+
+
+def clear_last_network_objects(net):
+    """Clears NEURON objects and saves the current Network instance
+    Parameters
+    ----------
+    net: an instance of Network
+        The current Network instance (context) from which NEURON objects will
+        be created
+    """
+    global last_network
+
+    if last_network is not None:
+        last_network._clear_neuron_objects()
+
+    net._clear_neuron_objects()
```

Comment:
shouldn't this go into `__exit__` ?

Reply:
After refactoring, I agree, this should go in `_neuron_network`'s `__exit__`.

Reply:
However, it is also needed in `build` in case the user tries to build multiple networks (e.g. `plot_simulate_evoked.py`)